clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -948,7 +948,7 @@ private synchronized void resetStateAndRejoin() {
     synchronized void resetGenerationOnResponseError(ApiKeys api, Errors error) {
         log.debug("Resetting generation after encountering {} from {} response and requesting re-join", error, api);
 
-        resetState();
+        resetStateAndRejoin();
```

Comment:
@dajac I feel this patch can fix https://issues.apache.org/jira/browse/KAFKA-8266

The test case in https://issues.apache.org/jira/browse/KAFKA-8266 restarts all brokers to activate new configs. However, the data of group (__consumer_offsets) may get lost if all brokers are killed too quick (the log folder is changed when restarting broker so the restarting broker has to fetch data from other brokers). The heartbeat of running consumer will encounter ```UNKNOWN_MEMBER_ID``` in sending heartbeat request since the group data is gone. Without this patch, the heartbeat thread is disabled and the state is in ```UNJOINED``` but the ```rejoinNeeded``` is still false. In short, the consumer is not going to rejoin group so we can't see expected error.

Reply:
Nice find! I analyzed the logs and I was actually trying to find out why the consumer was not rejoining. The logs coincides with what you described. It may be due to this as you pointed out.

Reply:
@chia7712 @dajac I think this bug is only introduced recently in this PR: https://github.com/apache/kafka/pull/8834. If your tests is executed before it then maybe there are other issues yet to be discovered.

Reply:
That test became flaky quite recently. Time wise, that could coincides. I will try to run the test without https://github.com/apache/kafka/pull/8834 to check tomorrow.

Reply:
It seems that we would rejoin on FENCED_INSTANCE_ID as well, is that intentional? https://github.com/apache/kafka/blob/bd462df20321ff5b75a7e3eae70634268582d90b/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java#L1106

Reply:
That is fine since in the outer caller `FENCED_INSTANCE_ID` would be treated as fatal, in line 1386.